### PR TITLE
[TOOLS-1185] Impersonate

### PR DESCRIPTION
**Pull request by jhonypg**

### CODE REVIEW

File: core/src/main/java/objective/taskboard/auth/AuthenticationService.java [hunk 1]
```diff
@@ -75,6 +79,34 @@ private TaskboardUser createTaskboardUser(String username) {
...
+
+        return new LoggedUserDetails(taskboardUser.getUsername(), roles, taskboardUser.isAdmin());
+    }
+
+    private void impersonate(LoggedUserDetails loggedInUser, String impersonateUsername) throws ImpersonateException {
+        if (isBlank(impersonateUsername))
+            throw ImpersonateException.blankHeaderValue();
+
+        jiraService.getJiraUserAsMaster(impersonateUsername)
+                .orElseThrow(() -> ImpersonateException.userToImpersonateNotFound(impersonateUsername));
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/auth/AuthenticationService.java]: MAJOR The return value of "orElseThrow" must be used. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2201)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/test/java/objective/taskboard/auth/LoggedUserDetailsTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        public LoggedUserDetailsBuilder impersonating(LoggedUserDetails user) {
+            this.impersonateUser = Optional.of(user);
+            return this;
+        }
+
+        public LoggedUserDetails build() {
+            LoggedUserDetails user = new LoggedUserDetails(username, jiraRoles, isAdmin);
+            if (impersonateUser.isPresent())
```
Comment (jytakano) [inline core/src/test/java/objective/taskboard/auth/LoggedUserDetailsTest.java]:
```suggestion
            impersonateUser.ifPresent(impersonate -> user.setImpersonateUser(impersonate));
```

File: core/src/main/java/objective/taskboard/jira/JiraService.java [hunk 3]
```diff
@@ -211,6 +211,19 @@ public JiraUser getJiraUser(String username) {
         return jiraEndpointAsUser.request(JiraUser.Service.class).get(username);
     }
 
+
+    @Cacheable(CacheConfiguration.JIRA_USER)
+    public Optional<JiraUser> getJiraUserAsMaster(String username) {
+        log.debug("⬣⬣⬣⬣⬣  getJiraUser");
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/jira/JiraService.java]: Done

File: core/src/main/java/objective/taskboard/auth/AuthenticationService.java [hunk 4]
```diff
@@ -75,6 +79,34 @@ private TaskboardUser createTaskboardUser(String username) {
...
+        List<JiraRole> roles = jiraService.getUserRoles(taskboardUser.getUsername()).stream()
+                .map(r -> new LoggedUserDetails.JiraRole(r.id, r.name, r.projectKey))
+                .collect(toList());
+
+        return new LoggedUserDetails(taskboardUser.getUsername(), roles, taskboardUser.isAdmin());
+    }
+
+    private void impersonate(LoggedUserDetails loggedInUser, String impersonateUsername) throws ImpersonateException {
+        if (isBlank(impersonateUsername))
+            throw ImpersonateException.blankHeaderValue();
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/auth/AuthenticationService.java]: I'll keep it.

File: core/src/main/java/objective/taskboard/auth/ImpersonateException.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+class ImpersonateException extends Exception {
+
+    private static final long serialVersionUID = 1L;
+
+    private ImpersonateException(String message) {
+        super(message);
+    }
+
+    public static ImpersonateException blankHeaderValue() {
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/auth/ImpersonateException.java]: I'll keep it.

File: core/src/main/java/objective/taskboard/auth/LoggedUserDetails.java [hunk 6]
```diff
@@ -1,29 +1,51 @@
...
-    public List<JiraRole> getJiraRoles() {
-        return jiraRoles;
+    public String getRealUsername() {
+        return username;
     }
 
-    public String getUsername() {
-        return username;
+    public Optional<String> getImpersonateUsername() {
+        return impersonateUser.isPresent()
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/auth/LoggedUserDetails.java]: Done

File: core/src/main/java/objective/taskboard/auth/LoggedUserDetails.java [hunk 7]
```diff
@@ -1,29 +1,51 @@
...
     }
-    
+
+    public String defineUsername() {
+        return getImpersonateUsername()
+                .orElse(getRealUsername());
+    }
+
+    public List<JiraRole> getJiraRoles() {
+        return impersonateUser.isPresent()
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/auth/LoggedUserDetails.java]: Done

File: core/src/main/java/objective/taskboard/auth/LoggedUserDetails.java [hunk 8]
```diff
@@ -1,29 +1,51 @@
...
+                ? impersonateUser.get().getJiraRoles()
+                : jiraRoles;
+    }
+
     public boolean isAdmin() {
-        return isAdmin;
+        return impersonateUser.isPresent()
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/auth/LoggedUserDetails.java]: Done

File: core/src/main/java/objective/taskboard/jira/endpoint/JiraEndpointAsLoggedInUser.java [hunk 9]
```diff
@@ -21,20 +19,44 @@
...
+        Headers.Builder headersBuilder = new Headers.Builder();
+
+        addImpersonateUser(headersBuilder);
+
+        return headersBuilder.build();
+    }
+
+    private void addImpersonateUser(Headers.Builder headersBuilder) {
+        Optional<String> impersonateUsername = CredentialsHolder.impersonateUsername();
+        if (impersonateUsername.isPresent())
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/main/java/objective/taskboard/jira/endpoint/JiraEndpointAsLoggedInUser.java]: Done

File: core/src/test/java/objective/taskboard/auth/AuthenticationServiceTest.java [hunk 10]
```diff
@@ -135,74 +255,110 @@ private RoleBuilder role() {
...
         }
 
-        public AuthenticationServiceTestDSL usernamePropertyToConnectWithJiraEquals(String propertyUsername) {
+        public AuthenticationServiceTestDSL usernamePropertyToConnectWithJira(String propertyUsername) {
             jiraProperties.getLousa().setUsername(propertyUsername);
             return this;
         }
 
+        /*
+        public AuthenticationServiceTestDSL jiraUserExists(String username) {
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/test/java/objective/taskboard/auth/AuthenticationServiceTest.java]: Done

File: core/src/test/java/objective/taskboard/auth/AuthenticationServiceTest.java [hunk 11]
```diff
@@ -41,85 +53,193 @@
...
+    @Test
+    public void authenticate_invalidCredentials_shouldFail() {
+        given()
+            .user("jose")
+                .doenstExistsOnJira()
+
+        .whenExecute(() -> authenticateResult = subject.authenticate("jose", "some-pass"))
+
+        .then()
+            .assertUser("mary")
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply core/src/test/java/objective/taskboard/auth/AuthenticationServiceTest.java]: Done

File: core/src/test/java/objective/taskboard/auth/LoggedUserDetailsTest.java (hunk 2 above)
Comment (jhonypg) [reply core/src/test/java/objective/taskboard/auth/LoggedUserDetailsTest.java]: Done

File: core/src/main/java/objective/taskboard/auth/AuthenticationService.java [hunk 12]
```diff
@@ -1,78 +1,98 @@
...
-            user.setAdmin(true);
+        loggedInUser.setImpersonateUser(getUserDetails(taskboardUserService.getTaskboardUser(impersonateUsername)));
+    }
 
-        taskboardUserRepository.add(user);
-        return user;
+    private void updateLastLogin(String username) {
+        try {
+            taskboardUserService.updateLastLoginToNow(username);
+        } catch (ObjectOptimisticLockingFailureException e) {
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/auth/AuthenticationService.java]: MAJOR Either log or rethrow this exception. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1166)

File: core/src/main/java/objective/taskboard/user/TaskboardUserService.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Transactional
+    public TaskboardUser getTaskboardUser(String username) {
+        return taskboardUserRepository.getByUsername(username)
+                .orElseGet(() -> createTaskboardUser(username));
+    }
+
+    @Transactional
+    private TaskboardUser createTaskboardUser(String username) {
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/user/TaskboardUserService.java]: MAJOR Make this method "public" or remove the "@Transactional" annotation [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2230)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**